File: 0007-reverse-integer/0007-reverse-integer.js
/**
 * @param {number} x
 * @return {number}
 */
// 1. Convert to string
// var reverse = function(x) {
//     const reversedStr = Math.abs(x).toString().split('').reverse().join("");
//     const result = Number(reversedStr) * Math.sign(x);

//      if (result < Math.pow(-2, 31) || result > Math.pow(2, 31) - 1) {
//         return 0;
//     }

//     return result;
// };

// 2. Using remainder
var reverse = function(x) {
    let num = Math.abs(x)
    let res = 0

    while (num > 0) {
        let digit = num % 10;
        num = Math.floor(num / 10);
        if (res > Math.floor((2 ** 31 - 1) / 10) || (res === Math.floor((2 ** 31 - 1) / 10) && digit > 7)) {
            return 0;
        }
        res = (res * 10) + digit;
    }

    return res * Math.sign(x)
}

File: 0007-reverse-integer/0007-reverse-integer.py
class Solution:
    def reverse(self, x: int) -> int:
        if x == 0:
            return 0
            
        reversed_str = str(abs(x))[::-1]
        result = int(reversed_str) * (1 if x > 0 else -1)

        if result < -2**31 or result > 2**31 - 1:
            return 0

        return result